Content.Server/PowerSink/PowerSinkSystem.cs

```diff
@@ -32,24 +52,82 @@ private void OnExamine(EntityUid uid, PowerSinkComponent component, ExaminedEven
 
         public override void Update(float frameTime)
         {
-            var toRemove = new RemQueue<(PowerSinkComponent Sink, BatteryComponent Battery)>();
+            var toRemove = new RemQueue<(EntityUid Entity, PowerSinkComponent Sink, BatteryComponent Battery)>();
+            var query = AllEntityQuery<PowerSinkComponent, PowerConsumerComponent, BatteryComponent, TransformComponent>();
 
             // Realistically it's gonna be like <5 per station.
-            foreach (var (comp, networkLoad, battery, xform) in EntityManager.EntityQuery<PowerSinkComponent, PowerConsumerComponent, BatteryComponent, TransformComponent>())
+            while (query.MoveNext(out var entity, out var component, out var networkLoad, out var battery, out var transform))
             {
-                if (!xform.Anchored) continue;
+                if (!transform.Anchored) continue;
 
                 battery.CurrentCharge += networkLoad.NetworkLoad.ReceivingPower / 1000;
+
+                var currentBatteryThreshold = battery.CurrentCharge / battery.MaxCharge;
+
+                // Check for warning message threshold
+                if (!component.SentImminentExplosionWarningMessage &&
+                    currentBatteryThreshold >= WarningMessageThreshold)
+                {
+                    NotifyStationOfImminentExplosion(entity, component);
+                }
+
+                // Check for warning sound threshold
+                foreach (var testThreshold in WarningSoundThresholds)
+                {
+                    if (currentBatteryThreshold >= testThreshold &&
+                        testThreshold > component.HighestWarningSoundThreshold)
+                    {
+                        component.HighestWarningSoundThreshold = currentBatteryThreshold; // Don't re-play in future until next threshold hit
+                        _audio.PlayPvs(component.ElectricSound, entity,
+                            new Robust.Shared.Audio.AudioParams(1f, 1f, "Master", SharedAudioSystem.DefaultSoundRange * 5, 1, 25, false, 0f) // audible from farther than usual
```

Comment:
The audioparams arg is just there for backwards compat atm, you can just change the data on ElectricSound itself to whatever you need (Params = AudioParams.Default etc).